jsii: interfaces incorrectly marked as datatypes

In the following case, the 

```ts
interface IBaseInterface {
  readonly value: string;
  doSomething(): void;
}

interface IDerivedInterface extends IBaseInterface {
  readonly otherValue: string;
}
```

The interface `IDerivedInterface` is incorrectly marked as `{datatype: true}`, because the interface consists only of properties. Correct behavior would have been to take the parent interfaces into account as well while making this assessment.

This leads the Java code generator to emit a Builder that contains duplicate fields (and setters) and ignoring the methods in the upstream classes (encountered while making an `IRole` interface):

```
@aws-cdk/aws-iam: [ERROR] /tmp/jsii-pacmak-codeUbHgJB/src/main/java/software/amazon/awscdk/services/iam/IRole.java:[26,69] variable _principal is already defined in class software.amazon.awscdk.services.iam.IRole.Builder
@aws-cdk/aws-iam: [ERROR] /tmp/jsii-pacmak-codeUbHgJB/src/main/java/software/amazon/awscdk/services/iam/IRole.java:[52,24] method withPrincipal(software.amazon.awscdk.services.iam.PolicyPrincipal) is already defined in class software.amazon.awscdk.services.iam.IRole.Builder
@aws-cdk/aws-iam: [ERROR] /tmp/jsii-pacmak-codeUbHgJB/src/main/java/software/amazon/awscdk/services/iam/IRole.java:[76,83] variable $principal is already defined in class <anonymous software.amazon.awscdk.services.iam.IRole$Builder$1>
@aws-cdk/aws-iam: [ERROR] /tmp/jsii-pacmak-codeUbHgJB/src/main/java/software/amazon/awscdk/services/iam/IRole.java:[90,76] method getPrincipal() is already defined in class <anonymous software.amazon.awscdk.services.iam.IRole$Builder$1>
@aws-cdk/aws-iam: [ERROR] /tmp/jsii-pacmak-codeUbHgJB/src/main/java/software/amazon/awscdk/services/iam/IRole.java:[73,32] <anonymous software.amazon.awscdk.services.iam.IRole$Builder$1> is not abstract and does not override abstract method attachManagedPolicy(java.lang.String) in software.amazon.awscdk.services.iam.IPrincipal
```

--- a/packages/jsii-calc/lib/compliance.ts
+++ b/packages/jsii-calc/lib/compliance.ts
@@ -922,3 +922,16 @@ export class ClassWithPrivateConstructorAndAutomaticProperties implements IInter
     private constructor(public readonly readOnlyString: string, public readWriteString: string) {
     }
 }
+
+export interface IInterfaceWithMethods {
+    readonly value: string;
+    doThings(): void;
+}
+
+/**
+ * Even though this interface has only properties, it is disqualified from being a datatype
+ * because it inherits from an interface that is not a datatype.
+ */
+export interface IInterfaceThatShouldNotBeADataType extends IInterfaceWithMethods {
+    readonly otherValue: string;
+}

--- a/packages/jsii-calc/test/assembly.jsii
+++ b/packages/jsii-calc/test/assembly.jsii
@@ -1429,6 +1429,52 @@
       "kind": "interface",
       "name": "IFriendlyRandomGenerator"
     },
+    "jsii-calc.IInterfaceThatShouldNotBeADataType": {
+      "assembly": "jsii-calc",
+      "docs": {
+        "comment": "Even though this interface has only properties, it is disqualified from being a datatype\nbecause it inherits from an interface that is not a datatype."
+      },
+      "fqn": "jsii-calc.IInterfaceThatShouldNotBeADataType",
+      "interfaces": [
+        {
+          "fqn": "jsii-calc.IInterfaceWithMethods"
+        }
+      ],
+      "kind": "interface",
+      "name": "IInterfaceThatShouldNotBeADataType",
+      "properties": [
+        {
+          "abstract": true,
+          "immutable": true,
+          "name": "otherValue",
+          "type": {
+            "primitive": "string"
+          }
+        }
+      ]
+    },
+    "jsii-calc.IInterfaceWithMethods": {
+      "assembly": "jsii-calc",
+      "fqn": "jsii-calc.IInterfaceWithMethods",
+      "kind": "interface",
+      "methods": [
+        {
+          "abstract": true,
+          "name": "doThings"
+        }
+      ],
+      "name": "IInterfaceWithMethods",
+      "properties": [
+        {
+          "abstract": true,
+          "immutable": true,
+          "name": "value",
+          "type": {
+            "primitive": "string"
+          }
+        }
+      ]
+    },
     "jsii-calc.IInterfaceWithProperties": {
       "assembly": "jsii-calc",
       "datatype": true,
@@ -3332,5 +3378,5 @@
     }
   },
   "version": "0.7.7",
-  "fingerprint": "SMX3FQtN8x0lUjxLt3hL50eC+4BhUchNcGPOAxkO9Xw="
+  "fingerprint": "bxhDOTLZBrkP11/RKcB3NDECWyzz3uPDRFgQXpHtrc0="
 }

--- a/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/.jsii
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/.jsii
@@ -1429,6 +1429,52 @@
       "kind": "interface",
       "name": "IFriendlyRandomGenerator"
     },
+    "jsii-calc.IInterfaceThatShouldNotBeADataType": {
+      "assembly": "jsii-calc",
+      "docs": {
+        "comment": "Even though this interface has only properties, it is disqualified from being a datatype\nbecause it inherits from an interface that is not a datatype."
+      },
+      "fqn": "jsii-calc.IInterfaceThatShouldNotBeADataType",
+      "interfaces": [
+        {
+          "fqn": "jsii-calc.IInterfaceWithMethods"
+        }
+      ],
+      "kind": "interface",
+      "name": "IInterfaceThatShouldNotBeADataType",
+      "properties": [
+        {
+          "abstract": true,
+          "immutable": true,
+          "name": "otherValue",
+          "type": {
+            "primitive": "string"
+          }
+        }
+      ]
+    },
+    "jsii-calc.IInterfaceWithMethods": {
+      "assembly": "jsii-calc",
+      "fqn": "jsii-calc.IInterfaceWithMethods",
+      "kind": "interface",
+      "methods": [
+        {
+          "abstract": true,
+          "name": "doThings"
+        }
+      ],
+      "name": "IInterfaceWithMethods",
+      "properties": [
+        {
+          "abstract": true,
+          "immutable": true,
+          "name": "value",
+          "type": {
+            "primitive": "string"
+          }
+        }
+      ]
+    },
     "jsii-calc.IInterfaceWithProperties": {
       "assembly": "jsii-calc",
       "datatype": true,
@@ -3332,5 +3378,5 @@
     }
   },
   "version": "0.7.7",
-  "fingerprint": "SMX3FQtN8x0lUjxLt3hL50eC+4BhUchNcGPOAxkO9Xw="
+  "fingerprint": "bxhDOTLZBrkP11/RKcB3NDECWyzz3uPDRFgQXpHtrc0="
 }

--- a/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IIInterfaceThatShouldNotBeADataType.cs
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IIInterfaceThatShouldNotBeADataType.cs
@@ -1,0 +1,18 @@
+using Amazon.JSII.Runtime.Deputy;
+
+namespace Amazon.JSII.Tests.CalculatorNamespace
+{
+    /// <summary>
+    /// Even though this interface has only properties, it is disqualified from being a datatype
+    /// because it inherits from an interface that is not a datatype.
+    /// </summary>
+    [JsiiInterface(typeof(IIInterfaceThatShouldNotBeADataType), "jsii-calc.IInterfaceThatShouldNotBeADataType")]
+    public interface IIInterfaceThatShouldNotBeADataType : IIInterfaceWithMethods
+    {
+        [JsiiProperty("otherValue", "{\"primitive\":\"string\"}")]
+        string OtherValue
+        {
+            get;
+        }
+    }
+}

--- a/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IIInterfaceWithMethods.cs
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IIInterfaceWithMethods.cs
@@ -1,0 +1,17 @@
+using Amazon.JSII.Runtime.Deputy;
+
+namespace Amazon.JSII.Tests.CalculatorNamespace
+{
+    [JsiiInterface(typeof(IIInterfaceWithMethods), "jsii-calc.IInterfaceWithMethods")]
+    public interface IIInterfaceWithMethods
+    {
+        [JsiiProperty("value", "{\"primitive\":\"string\"}")]
+        string Value
+        {
+            get;
+        }
+
+        [JsiiMethod("doThings", null, "[]")]
+        void DoThings();
+    }
+}

--- a/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IInterfaceThatShouldNotBeADataTypeProxy.cs
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IInterfaceThatShouldNotBeADataTypeProxy.cs
@@ -1,0 +1,34 @@
+using Amazon.JSII.Runtime.Deputy;
+
+namespace Amazon.JSII.Tests.CalculatorNamespace
+{
+    /// <summary>
+    /// Even though this interface has only properties, it is disqualified from being a datatype
+    /// because it inherits from an interface that is not a datatype.
+    /// </summary>
+    [JsiiTypeProxy(typeof(IIInterfaceThatShouldNotBeADataType), "jsii-calc.IInterfaceThatShouldNotBeADataType")]
+    internal sealed class IInterfaceThatShouldNotBeADataTypeProxy : DeputyBase, IIInterfaceThatShouldNotBeADataType
+    {
+        private IInterfaceThatShouldNotBeADataTypeProxy(ByRefValue reference): base(reference)
+        {
+        }
+
+        [JsiiProperty("otherValue", "{\"primitive\":\"string\"}")]
+        public string OtherValue
+        {
+            get => GetInstanceProperty<string>();
+        }
+
+        [JsiiProperty("value", "{\"primitive\":\"string\"}")]
+        public string Value
+        {
+            get => GetInstanceProperty<string>();
+        }
+
+        [JsiiMethod("doThings", null, "[]")]
+        public void DoThings()
+        {
+            InvokeInstanceVoidMethod(new object[]{});
+        }
+    }
+}

--- a/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IInterfaceWithMethodsProxy.cs
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/dotnet/Amazon.JSII.Tests.CalculatorPackageId/Amazon/JSII/Tests/CalculatorNamespace/IInterfaceWithMethodsProxy.cs
@@ -1,0 +1,24 @@
+using Amazon.JSII.Runtime.Deputy;
+
+namespace Amazon.JSII.Tests.CalculatorNamespace
+{
+    [JsiiTypeProxy(typeof(IIInterfaceWithMethods), "jsii-calc.IInterfaceWithMethods")]
+    internal sealed class IInterfaceWithMethodsProxy : DeputyBase, IIInterfaceWithMethods
+    {
+        private IInterfaceWithMethodsProxy(ByRefValue reference): base(reference)
+        {
+        }
+
+        [JsiiProperty("value", "{\"primitive\":\"string\"}")]
+        public string Value
+        {
+            get => GetInstanceProperty<string>();
+        }
+
+        [JsiiMethod("doThings", null, "[]")]
+        public void DoThings()
+        {
+            InvokeInstanceVoidMethod(new object[]{});
+        }
+    }
+}

--- a/packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/$Module.java
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/$Module.java
@@ -40,6 +40,8 @@ public final class $Module extends JsiiModule {
             case "jsii-calc.GiveMeStructs": return software.amazon.jsii.tests.calculator.GiveMeStructs.class;
             case "jsii-calc.IFriendlier": return software.amazon.jsii.tests.calculator.IFriendlier.class;
             case "jsii-calc.IFriendlyRandomGenerator": return software.amazon.jsii.tests.calculator.IFriendlyRandomGenerator.class;
+            case "jsii-calc.IInterfaceThatShouldNotBeADataType": return software.amazon.jsii.tests.calculator.IInterfaceThatShouldNotBeADataType.class;
+            case "jsii-calc.IInterfaceWithMethods": return software.amazon.jsii.tests.calculator.IInterfaceWithMethods.class;
             case "jsii-calc.IInterfaceWithProperties": return software.amazon.jsii.tests.calculator.IInterfaceWithProperties.class;
             case "jsii-calc.IInterfaceWithPropertiesExtension": return software.amazon.jsii.tests.calculator.IInterfaceWithPropertiesExtension.class;
             case "jsii-calc.IRandomNumberGenerator": return software.amazon.jsii.tests.calculator.IRandomNumberGenerator.class;

--- a/packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/IInterfaceThatShouldNotBeADataType.java
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/IInterfaceThatShouldNotBeADataType.java
@@ -1,0 +1,34 @@
+package software.amazon.jsii.tests.calculator;
+
+/**
+ * Even though this interface has only properties, it is disqualified from being a datatype
+ * because it inherits from an interface that is not a datatype.
+ */
+@javax.annotation.Generated(value = "jsii-pacmak")
+public interface IInterfaceThatShouldNotBeADataType extends software.amazon.jsii.JsiiSerializable, software.amazon.jsii.tests.calculator.IInterfaceWithMethods {
+    java.lang.String getOtherValue();
+
+    /**
+     * A proxy class which represents a concrete javascript instance of this type.
+     */
+    final static class Jsii$Proxy extends software.amazon.jsii.JsiiObject implements software.amazon.jsii.tests.calculator.IInterfaceThatShouldNotBeADataType {
+        protected Jsii$Proxy(final software.amazon.jsii.JsiiObject.InitializationMode mode) {
+            super(mode);
+        }
+
+        @Override
+        public java.lang.String getOtherValue() {
+            return this.jsiiGet("otherValue", java.lang.String.class);
+        }
+
+        @Override
+        public java.lang.String getValue() {
+            return this.jsiiGet("value", java.lang.String.class);
+        }
+
+        @Override
+        public void doThings() {
+            this.jsiiCall("doThings", Void.class);
+        }
+    }
+}

--- a/packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/IInterfaceWithMethods.java
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/IInterfaceWithMethods.java
@@ -1,0 +1,26 @@
+package software.amazon.jsii.tests.calculator;
+
+@javax.annotation.Generated(value = "jsii-pacmak")
+public interface IInterfaceWithMethods extends software.amazon.jsii.JsiiSerializable {
+    java.lang.String getValue();
+    void doThings();
+
+    /**
+     * A proxy class which represents a concrete javascript instance of this type.
+     */
+    final static class Jsii$Proxy extends software.amazon.jsii.JsiiObject implements software.amazon.jsii.tests.calculator.IInterfaceWithMethods {
+        protected Jsii$Proxy(final software.amazon.jsii.JsiiObject.InitializationMode mode) {
+            super(mode);
+        }
+
+        @Override
+        public java.lang.String getValue() {
+            return this.jsiiGet("value", java.lang.String.class);
+        }
+
+        @Override
+        public void doThings() {
+            this.jsiiCall("doThings", Void.class);
+        }
+    }
+}

--- a/packages/jsii-pacmak/test/expected.jsii-calc/sphinx/jsii-calc.rst
+++ b/packages/jsii-pacmak/test/expected.jsii-calc/sphinx/jsii-calc.rst
@@ -1547,6 +1547,97 @@ IFriendlyRandomGenerator (interface)
       :abstract: Yes
 
 
+IInterfaceThatShouldNotBeADataType (interface)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. py:class:: IInterfaceThatShouldNotBeADataType
+
+   **Language-specific names:**
+
+   .. tabs::
+
+      .. code-tab:: c#
+
+         using Amazon.JSII.Tests.CalculatorNamespace;
+
+      .. code-tab:: java
+
+         import software.amazon.jsii.tests.calculator.IInterfaceThatShouldNotBeADataType;
+
+      .. code-tab:: javascript
+
+         // IInterfaceThatShouldNotBeADataType is an interface
+
+      .. code-tab:: typescript
+
+         import { IInterfaceThatShouldNotBeADataType } from 'jsii-calc';
+
+
+
+   Even though this interface has only properties, it is disqualified from being a datatype because it inherits from an interface that is not a datatype.
+
+
+   :extends: :py:class:`~jsii-calc.IInterfaceWithMethods`\ 
+
+
+   .. py:attribute:: otherValue
+
+      :type: string *(readonly)* *(abstract)*
+
+
+   .. py:method:: doThings()
+
+      *Inherited from* :py:meth:`jsii-calc.IInterfaceWithMethods <jsii-calc.IInterfaceWithMethods.doThings>`
+
+      :abstract: Yes
+
+
+   .. py:attribute:: value
+
+      *Inherited from* :py:attr:`jsii-calc.IInterfaceWithMethods <jsii-calc.IInterfaceWithMethods.value>`
+
+      :type: string *(readonly)* *(abstract)*
+
+
+IInterfaceWithMethods (interface)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. py:class:: IInterfaceWithMethods
+
+   **Language-specific names:**
+
+   .. tabs::
+
+      .. code-tab:: c#
+
+         using Amazon.JSII.Tests.CalculatorNamespace;
+
+      .. code-tab:: java
+
+         import software.amazon.jsii.tests.calculator.IInterfaceWithMethods;
+
+      .. code-tab:: javascript
+
+         // IInterfaceWithMethods is an interface
+
+      .. code-tab:: typescript
+
+         import { IInterfaceWithMethods } from 'jsii-calc';
+
+
+
+
+
+   .. py:attribute:: value
+
+      :type: string *(readonly)* *(abstract)*
+
+
+   .. py:method:: doThings()
+
+      :abstract: Yes
+
+
 IInterfaceWithProperties (interface)
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 

--- a/packages/jsii/lib/assembler.ts
+++ b/packages/jsii/lib/assembler.ts
@@ -11,6 +11,7 @@ import literate = require('./literate');
 import { ProjectInfo } from './project-info';
 import utils = require('./utils');
 import { Validator } from './validator';
+import { NamedTypeReference, isInterfaceType } from 'jsii-spec';
 
 // tslint:disable:no-var-requires Modules without TypeScript definitions
 const sortJson = require('sort-json');
@@ -23,7 +24,7 @@ const LOG = log4js.getLogger('jsii/assembler');
  */
 export class Assembler implements Emitter {
     private _diagnostics = new Array<Diagnostic>();
-    private _deferred = new Array<() => void>();
+    private _deferred = new Array<DeferredRecord>();
     private _types: { [fqn: string]: spec.Type };
 
     /**
@@ -75,9 +76,8 @@ export class Assembler implements Emitter {
                 await this._visitNode(node.declarations[0]);
             }
         }
-        for (const deferred of this._deferred) {
-            deferred();
-        }
+
+        this.callDeferredsInOrder();
 
         // Skip emitting if any diagnostic message is an error
         if (this._diagnostics.find(diag => diag.category === ts.DiagnosticCategory.Error) != null) {
@@ -87,9 +87,8 @@ export class Assembler implements Emitter {
             try {
                 return { diagnostics: this._diagnostics, emitSkipped: true };
             } finally {
-                // Clearing ``this._diagnostics`` and ``this._deferred`` to allow contents to be garbage-collected.
+                // Clearing ``this._diagnostics`` to allow contents to be garbage-collected.
                 delete this._diagnostics;
-                delete this._deferred;
             }
         }
 
@@ -128,9 +127,8 @@ export class Assembler implements Emitter {
             // Clearing ``this._types`` to allow contents to be garbage-collected.
             delete this._types;
 
-            // Clearing ``this._diagnostics`` and ``this._deferred`` to allow contents to be garbage-collected.
+            // Clearing ``this._diagnostics`` to allow contents to be garbage-collected.
             delete this._diagnostics;
-            delete this._deferred;
         }
 
         async function _loadReadme(this: Assembler) {
@@ -148,11 +146,38 @@ export class Assembler implements Emitter {
      * Defer checks for after the program has been entirely processed; useful for verifying type references that may not
      * have been discovered yet, and verifying properties about them.
      *
+     * @param fqn FQN of the current type.
+     * @param deps List of FQNs of types this callback depends on. All deferreds for all
      * @param cb the function to be called in a deferred way. It will be bound with ``this``, so it can depend on using
      *           ``this``.
      */
-    private _defer(cb: () => void) {
-        this._deferred.push(cb.bind(this));
+    private _defer(fqn: string, depFqns: string[], cb: () => void) {
+        this._deferred.push({ fqn, depFqns, cb: cb.bind(this) });
+    }
+
+    /**
+     * Defer a callback until a (set of) types are available
+     *
+     * This is a helper function around _defer() which encapsulates the _dereference
+     * action (which is basically the majority use case for _defer anyway).
+     *
+     * Will not invoke the function with any 'undefined's; an error will already have been emitted in
+     * that case anyway.
+     */
+    // tslint:disable-next-line:max-line-length
+    private _deferUntilTypeAvailable(fqn: string, baseTypes: NamedTypeReference[], referencingNode: ts.Node, cb: (...xs: spec.Type[]) => void) {
+        // We can do this one eagerly
+        if (baseTypes.length === 0) {
+            cb();
+            return;
+        }
+
+        this._defer(fqn, baseTypes.map(x => x.fqn), () => {
+            const resolved = baseTypes.map(x => this._dereference(x, referencingNode)).filter(x => x !== undefined);
+            if (resolved.length > 0) {
+                return cb(...resolved as spec.Type[]);
+            }
+        });
     }
 
     /**
@@ -299,9 +324,11 @@ export class Assembler implements Emitter {
             LOG.trace(`Processing class: ${colors.gray(namespace.join('.'))}.${colors.cyan(type.symbol.name)}`);
         }
 
+        const fqn = `${[this.projectInfo.name, ...namespace].join('.')}.${type.symbol.name}`;
+
         const jsiiType: spec.ClassType = {
             assembly: this.projectInfo.name,
-            fqn: `${[this.projectInfo.name, ...namespace].join('.')}.${type.symbol.name}`,
+            fqn,
             kind: spec.TypeKind.Class,
             name: type.symbol.name,
             namespace: namespace.join('.')
@@ -322,9 +349,8 @@ export class Assembler implements Emitter {
                                 `Base type of ${jsiiType.fqn} is not a named type (${spec.describeTypeReference(ref)})`);
                 continue;
             }
-            this._defer(() => {
-                const deref = this._dereference(ref, base.symbol.valueDeclaration);
-                if (deref && !spec.isClassType(deref)) {
+            this._deferUntilTypeAvailable(fqn, [ref], base.symbol.valueDeclaration, (deref) => {
+                if (!spec.isClassType(deref)) {
                     this._diagnostic(base.symbol.valueDeclaration,
                                     ts.DiagnosticCategory.Error,
                                     `Base type of ${jsiiType.fqn} is not a class (${spec.describeTypeReference(ref)})`);
@@ -349,9 +375,8 @@ export class Assembler implements Emitter {
                                      `Interface of ${jsiiType.fqn} is not a named type (${spec.describeTypeReference(typeRef)})`);
                     continue;
                 }
-                this._defer(() => {
-                    const deref = this._dereference(typeRef, expression);
-                    if (deref && !spec.isInterfaceType(deref)) {
+                this._deferUntilTypeAvailable(fqn, [typeRef], expression, (deref) => {
+                    if (!spec.isInterfaceType(deref)) {
                         this._diagnostic(expression,
                                         ts.DiagnosticCategory.Error,
                                         `Implements clause of ${jsiiType.fqn} uses ${spec.describeTypeReference(typeRef)} as an interface`);
@@ -410,16 +435,13 @@ export class Assembler implements Emitter {
                 }
             }
         } else if (jsiiType.base) {
-            this._defer(() => {
-                const baseType = this._dereference(jsiiType.base!, type.symbol.valueDeclaration);
-                if (baseType) {
-                    if (spec.isClassType(baseType)) {
-                        jsiiType.initializer = baseType.initializer;
-                    } else {
-                        this._diagnostic(type.symbol.valueDeclaration,
-                            ts.DiagnosticCategory.Error,
-                            `Base type of ${jsiiType.fqn} (${jsiiType.base!.fqn}) is not a class`);
-                    }
+            this._deferUntilTypeAvailable(fqn, [jsiiType.base!], type.symbol.valueDeclaration, (baseType) => {
+                if (spec.isClassType(baseType)) {
+                    jsiiType.initializer = baseType.initializer;
+                } else {
+                    this._diagnostic(type.symbol.valueDeclaration,
+                        ts.DiagnosticCategory.Error,
+                        `Base type of ${jsiiType.fqn} (${jsiiType.base!.fqn}) is not a class`);
                 }
             });
         } else {
@@ -499,13 +521,16 @@ export class Assembler implements Emitter {
             LOG.trace(`Processing interface: ${colors.gray(namespace.join('.'))}.${colors.cyan(type.symbol.name)}`);
         }
 
+        const fqn = `${[this.projectInfo.name, ...namespace].join('.')}.${type.symbol.name}`;
+
         const jsiiType: spec.InterfaceType = {
             assembly: this.projectInfo.name,
-            fqn: `${[this.projectInfo.name, ...namespace].join('.')}.${type.symbol.name}`,
+            fqn,
             kind: spec.TypeKind.Interface,
             name: type.symbol.name,
             namespace: namespace.join('.')
         };
+
         for (const base of (type.getBaseTypes() || [])) {
             const ref = await this._typeReference(base, type.symbol.valueDeclaration);
             if (!spec.isNamedTypeReference(ref)) {
@@ -514,9 +539,8 @@ export class Assembler implements Emitter {
                                  `Base type of ${jsiiType.fqn} is not a named type (${spec.describeTypeReference(ref)})`);
                 continue;
             }
-            this._defer(() => {
-                const baseType = this._dereference(ref, base.symbol.valueDeclaration);
-                if (baseType && !spec.isInterfaceType(baseType)) {
+            this._deferUntilTypeAvailable(fqn, [ref], base.symbol.valueDeclaration, (baseType) => {
+                if (!spec.isInterfaceType(baseType)) {
                     // tslint:disable:max-line-length
                     this._diagnostic(base.symbol.valueDeclaration,
                                     ts.DiagnosticCategory.Error,
@@ -545,9 +569,20 @@ export class Assembler implements Emitter {
                                  `Ignoring un-handled ${ts.SyntaxKind[member.valueDeclaration.kind]} member`);
             }
         }
-        if ((jsiiType.methods || []).length === 0 && (jsiiType.properties || []).length > 0) {
-            jsiiType.datatype = true;
-        }
+
+        // Calculate datatype based on the datatypeness of this interface and all of its parents
+        // To keep the spec minimal the actual values of the attribute are "true" or "undefined" (to represent "false").
+        // tslint:disable-next-line:no-console
+        this._deferUntilTypeAvailable(fqn, jsiiType.interfaces || [], type.symbol.valueDeclaration, (...bases: spec.Type[]) => {
+            if ((jsiiType.methods || []).length === 0 && (jsiiType.properties || []).length > 0) {
+                jsiiType.datatype = true;
+            }
+            for (const base of bases) {
+                if (isInterfaceType(base) && !base.datatype) {
+                    jsiiType.datatype = undefined;
+                }
+            }
+        });
 
         return _sortMembers(this._visitDocumentation(type.symbol, jsiiType));
     }
@@ -780,6 +815,41 @@ export class Assembler implements Emitter {
                 : { union: { types }, optional };
         }
     }
+
+    private callDeferredsInOrder() {
+        // Do a topological call order of all deferreds.
+        while (this._deferred.length > 0) {
+            // All fqns in dependency lists that don't have any pending
+            // deferreds themselves can be executed now, so are removed from
+            // dependency lists.
+            const pendingFqns = new Set<string>(this._deferred.map(x => x.fqn));
+            for (const deferred of this._deferred) {
+                restrictDependenciesTo(deferred, pendingFqns);
+            }
+
+            // Invoke all deferreds with no more dependencies and remove them from the list.
+            let invoked = false;
+            for (let i = 0; i < this._deferred.length; i++) {
+                if (this._deferred[i].depFqns.length === 0) {
+                    const deferred = this._deferred.splice(i, 1)[0];
+                    deferred.cb();
+                    invoked = true;
+                }
+            }
+
+            if (!invoked) {
+                // Apparently we're stuck. Complain loudly.
+                throw new Error(`Could not invoke any more deferreds, cyclic dependency? Remaining: ${JSON.stringify(this._deferred, undefined, 2)}`);
+            }
+        }
+
+        /**
+         * Retain only elements in the dependencyfqn that are also in the set
+         */
+        function restrictDependenciesTo(def: DeferredRecord, fqns: Set<string>) {
+            def.depFqns = def.depFqns.filter(fqns.has.bind(fqns));
+        }
+    }
 }
 
 /**
@@ -912,4 +982,10 @@ function _toDependencies(assemblies: ReadonlyArray<spec.Assembly>): { [name: str
         };
     }
     return result;
+}
+
+interface DeferredRecord {
+    fqn: string;
+    depFqns: string[];
+    cb: () => void;
 }

--- a/packages/jsii/lib/assembler.ts
+++ b/packages/jsii/lib/assembler.ts
@@ -143,19 +143,6 @@ export class Assembler implements Emitter {
     }
 
     /**
-     * Defer checks for after the program has been entirely processed; useful for verifying type references that may not
-     * have been discovered yet, and verifying properties about them.
-     *
-     * @param fqn FQN of the current type.
-     * @param deps List of FQNs of types this callback depends on. All deferreds for all
-     * @param cb the function to be called in a deferred way. It will be bound with ``this``, so it can depend on using
-     *           ``this``.
-     */
-    private _defer(fqn: string, depFqns: string[], cb: () => void) {
-        this._deferred.push({ fqn, depFqns, cb: cb.bind(this) });
-    }
-
-    /**
      * Defer a callback until a (set of) types are available
      *
      * This is a helper function around _defer() which encapsulates the _dereference
@@ -165,7 +152,7 @@ export class Assembler implements Emitter {
      * that case anyway.
      */
     // tslint:disable-next-line:max-line-length
-    private _deferUntilTypeAvailable(fqn: string, baseTypes: NamedTypeReference[], referencingNode: ts.Node, cb: (...xs: spec.Type[]) => void) {
+    private _deferUntilTypesAvailable(fqn: string, baseTypes: NamedTypeReference[], referencingNode: ts.Node, cb: (...xs: spec.Type[]) => void) {
         // We can do this one eagerly
         if (baseTypes.length === 0) {
             cb();
@@ -178,6 +165,22 @@ export class Assembler implements Emitter {
                 return cb(...resolved as spec.Type[]);
             }
         });
+    }
+
+    /**
+     * Defer checks for after the program has been entirely processed; useful for verifying type references that may not
+     * have been discovered yet, and verifying properties about them.
+     *
+     * The callback is guaranteed to be executed only after all deferreds for all types in 'dependedFqns' have
+     * been executed.
+     *
+     * @param fqn FQN of the current type.
+     * @param deps List of FQNs of types this callback depends on. All deferreds for all
+     * @param cb the function to be called in a deferred way. It will be bound with ``this``, so it can depend on using
+     *           ``this``.
+     */
+    private _defer(fqn: string, dependedFqns: string[], cb: () => void) {
+        this._deferred.push({ fqn, dependedFqns, cb: cb.bind(this) });
     }
 
     /**
@@ -349,7 +352,7 @@ export class Assembler implements Emitter {
                                 `Base type of ${jsiiType.fqn} is not a named type (${spec.describeTypeReference(ref)})`);
                 continue;
             }
-            this._deferUntilTypeAvailable(fqn, [ref], base.symbol.valueDeclaration, (deref) => {
+            this._deferUntilTypesAvailable(fqn, [ref], base.symbol.valueDeclaration, (deref) => {
                 if (!spec.isClassType(deref)) {
                     this._diagnostic(base.symbol.valueDeclaration,
                                     ts.DiagnosticCategory.Error,
@@ -375,7 +378,7 @@ export class Assembler implements Emitter {
                                      `Interface of ${jsiiType.fqn} is not a named type (${spec.describeTypeReference(typeRef)})`);
                     continue;
                 }
-                this._deferUntilTypeAvailable(fqn, [typeRef], expression, (deref) => {
+                this._deferUntilTypesAvailable(fqn, [typeRef], expression, (deref) => {
                     if (!spec.isInterfaceType(deref)) {
                         this._diagnostic(expression,
                                         ts.DiagnosticCategory.Error,
@@ -435,7 +438,7 @@ export class Assembler implements Emitter {
                 }
             }
         } else if (jsiiType.base) {
-            this._deferUntilTypeAvailable(fqn, [jsiiType.base!], type.symbol.valueDeclaration, (baseType) => {
+            this._deferUntilTypesAvailable(fqn, [jsiiType.base!], type.symbol.valueDeclaration, (baseType) => {
                 if (spec.isClassType(baseType)) {
                     jsiiType.initializer = baseType.initializer;
                 } else {
@@ -539,7 +542,7 @@ export class Assembler implements Emitter {
                                  `Base type of ${jsiiType.fqn} is not a named type (${spec.describeTypeReference(ref)})`);
                 continue;
             }
-            this._deferUntilTypeAvailable(fqn, [ref], base.symbol.valueDeclaration, (baseType) => {
+            this._deferUntilTypesAvailable(fqn, [ref], base.symbol.valueDeclaration, (baseType) => {
                 if (!spec.isInterfaceType(baseType)) {
                     // tslint:disable:max-line-length
                     this._diagnostic(base.symbol.valueDeclaration,
@@ -573,7 +576,7 @@ export class Assembler implements Emitter {
         // Calculate datatype based on the datatypeness of this interface and all of its parents
         // To keep the spec minimal the actual values of the attribute are "true" or "undefined" (to represent "false").
         // tslint:disable-next-line:no-console
-        this._deferUntilTypeAvailable(fqn, jsiiType.interfaces || [], type.symbol.valueDeclaration, (...bases: spec.Type[]) => {
+        this._deferUntilTypesAvailable(fqn, jsiiType.interfaces || [], type.symbol.valueDeclaration, (...bases: spec.Type[]) => {
             if ((jsiiType.methods || []).length === 0 && (jsiiType.properties || []).length > 0) {
                 jsiiType.datatype = true;
             }
@@ -830,7 +833,7 @@ export class Assembler implements Emitter {
             // Invoke all deferreds with no more dependencies and remove them from the list.
             let invoked = false;
             for (let i = 0; i < this._deferred.length; i++) {
-                if (this._deferred[i].depFqns.length === 0) {
+                if (this._deferred[i].dependedFqns.length === 0) {
                     const deferred = this._deferred.splice(i, 1)[0];
                     deferred.cb();
                     invoked = true;
@@ -847,7 +850,7 @@ export class Assembler implements Emitter {
          * Retain only elements in the dependencyfqn that are also in the set
          */
         function restrictDependenciesTo(def: DeferredRecord, fqns: Set<string>) {
-            def.depFqns = def.depFqns.filter(fqns.has.bind(fqns));
+            def.dependedFqns = def.dependedFqns.filter(fqns.has.bind(fqns));
         }
     }
 }
@@ -984,8 +987,25 @@ function _toDependencies(assemblies: ReadonlyArray<spec.Assembly>): { [name: str
     return result;
 }
 
+/**
+ * Deferred processing that needs to happen in a second, ordered pass
+ */
 interface DeferredRecord {
+    /**
+     * The FQN of the type the action will be executed on
+     */
     fqn: string;
-    depFqns: string[];
+
+    /**
+     * Dependency FQNs of the types that need to be processed before analysis.
+     *
+     * All deferred analysis actions for the types listed here must be complete
+     * before this analysis action can run.
+     */
+    dependedFqns: string[];
+
+    /**
+     * Callback representing the action to run.
+     */
     cb: () => void;
 }

